core/modules/shrscan/Director.h

```diff
@@ -0,0 +1,315 @@
+
+class Director
+{
+public:
+
+//-----------------------------------------------------------------------------
+//! @brief Add a query to the shared scan queue.
+//!
+//! @param  query - Pointer to a Query object that describes the query.
+//-----------------------------------------------------------------------------
+
+virtual void  addQuery(std::shared_ptr<Query> const& query) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Cancel a shared scan query.
+//!
+//! When a query is cancelled, the Query object's endQuery() method is called.
+//! If the query is executing its runQuery() method, cancellation occurs upon
+//! return.
+//!
+//! @param  query   - Pointer to a Query object that describes the query.
+//! @param  wait    - When true, does not return until the query has been
+//!                   actually cancelled (i.e. endQuery() method called).
+//!
+//! @return false: The query was not found.
+//! @return true:  The query was cancelled or was scheduled to be cancelled.
+//-----------------------------------------------------------------------------
+
+virtual bool  cancelQuery(Query const* query, bool wait=false) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Cancel all shared scan queries.
+//!
+//! @param  wait    - When true, does not return until all queries have been
+//!                   actually cancelled (i.e. endQuery() method called).
+//!
+//! This method effectively calls cancelQuery() on each query.
+//-----------------------------------------------------------------------------
+
+virtual void  cancelAll(bool wait=false) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Obtain statistics about this shared scan director.
+//!
+//! @param  stats   - Where information is to be returned (see struct below).
+//!
+//! @return The number of queries actually running.
+//-----------------------------------------------------------------------------
+
+struct Statistics {
+    unsigned long long bytesLockMax;//< Maximum number of bytes to lock
+    unsigned long long bytesLocked; //< Actual  number of bytes locked in memory
+    unsigned long long msRunTotal;  //< Total milliseconds spent running
+    unsigned long long msIdleTotal; //< Total milliseconds spent idling
+    unsigned int       numChunks;   //< The number of chunks being handled
+    unsigned int       maxCanRun;   //< Maximum number of parallel queries
+    unsigned int       numQueued;   //< Number waiting to run
+    unsigned int       numRunning;  //< Number actually running
+    unsigned int       numSuspend;  //< Number suspended
+    unsigned int       numErrors;   //< Number of queries that had errors
+    unsigned int       numCancelled;//< Number of queries that were cancelled
+};
+
+virtual void  getStatistics(Statistics &stats) = 0;
```

Comment:
It does not solve a problem because there is no problem. If you want to copy something from mutex-protected data then one copy has to happen anyways. And there is no reason to copy it more than once no matter how you do it, NRVO is not going to be disabled just because you have a mutex somewhere.